File: data-structures/circular-queue/CircularQueue.py
class CircularQueue:
    def __init__(self, size):
        self.size = size
        self.queue = [None] * size
        self.front = -1
        self.rear = -1

    def isFull(self):
        return (self.rear + 1) % self.size == self.front

    def isEmpty(self):
        return self.front == -1

    def enqueue(self, value):
        if self.isFull():
            print("Queue is full! Cannot enqueue.")
            return
        if self.isEmpty():
            self.front = self.rear = 0
        else:
            self.rear = (self.rear + 1) % self.size
        self.queue[self.rear] = value
        print(f"Enqueued: {value}")

    def dequeue(self):
        if self.isEmpty():
            print("Queue is empty! Cannot dequeue.")
            return None
        removed = self.queue[self.front]
        if self.front == self.rear:
            self.front = self.rear = -1
        else:
            self.front = (self.front + 1) % self.size
        print(f"Dequeued: {removed}")
        return removed

    def peek(self):
        if self.isEmpty():
            print("Queue is empty!")
            return None
        return self.queue[self.front]

    def display(self):
        if self.isEmpty():
            print("Queue is empty.")
            return
        print("Queue contents:", end=" ")
        i = self.front
        while True:
            print(self.queue[i], end=" ")
            if i == self.rear:
                break
            i = (i + 1) % self.size
        print()

File: data-structures/circular-queue/CircularQueue.java

public class CircularQueue {
  private int[] arr;
  private int maxSize, currSize;
  private int f, b;

  public CircularQueue(int size) {
    maxSize = size;
    arr = new int[maxSize];
    currSize = 0;
    f = 0;
    b = -1;
  }

  public void enqueue(int value) {
    if (currSize == maxSize) {
      System.out.println("circular queue is full!");
      return;
    }
    b = (b + 1) % maxSize;
    arr[b] = value;
    currSize++;
  }

  public void dequeue() {
    if (isEmpty()) {
      System.out.println("circular queue is empty!");
      return;
    }
    f = (f + 1) % maxSize;
    currSize--;
  }

  public int front() {
    if (isEmpty()) {
      System.out.println("circular queue is empty!");
      return -1;
    }
    return arr[f];
  }


  public int last() {
    if (isEmpty()) {
      System.out.println("circular queue is empty!");
      return -1;
    }
    return arr[b];
  }

  public boolean isFull() {
    return currSize == maxSize;
  }

  public boolean isEmpty() {
    return currSize == 0;
  }

  public int size() {
    return currSize;
  }
}
